### PR TITLE
Store session keys by JoinEUI

### PR DESCRIPTION
**Pull request by johanstokking**
#### Summary

Closes https://github.com/TheThingsIndustries/lorawan-stack/issues/1826
References https://github.com/TheThingsIndustries/lorawan-stack/issues/1825

#### Changes

- Index JS session keys by `JoinEUI` too
- Make JS device registry context-aware

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/joinserver/registry.go [hunk 1]
```diff
@@ -23,9 +23,9 @@ import (
 
 // DeviceRegistry is a registry, containing devices.
 type DeviceRegistry interface {
-	GetByEUI(ctx context.Context, joinEUI types.EUI64, devEUI types.EUI64, paths []string) (*ttnpb.EndDevice, error)
+	GetByEUI(ctx context.Context, joinEUI types.EUI64, devEUI types.EUI64, paths []string) (*ttnpb.ContextualEndDevice, error)
 	GetByID(ctx context.Context, appID ttnpb.ApplicationIdentifiers, devID string, paths []string) (*ttnpb.EndDevice, error)
-	SetByEUI(ctx context.Context, joinEUI types.EUI64, devEUI types.EUI64, paths []string, f func(*ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.EndDevice, error)
+	SetByEUI(ctx context.Context, joinEUI types.EUI64, devEUI types.EUI64, paths []string, f func(*ttnpb.ContextualEndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.ContextualEndDevice, error)
```
Comment (rvolosatovs) [inline pkg/joinserver/registry.go]: IMO this is cleaner:
```suggestion
func (r *DeviceRegistry) SetByEUI(ctx context.Context, joinEUI types.EUI64, devEUI types.EUI64, gets []string, f func(ctx context.Context, pb *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.ContextualEndDevice, error) {
```

File: pkg/ttnpb/contextual_end_device.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttnpb
+
+import "context"
+
+// ContextualEndDevice is an end device with its context.
+type ContextualEndDevice struct {
```
Comment (rvolosatovs) [inline pkg/ttnpb/contextual_end_device.go]: Since it's only used in JS, I think it's fine to put it there

Comment (johanstokking) [reply pkg/ttnpb/contextual_end_device.go]: We get import cycles then; `pkg/joinserver/redis` can't import `pkg/joinserver`

Comment (rvolosatovs) [reply pkg/ttnpb/contextual_end_device.go]: That should not be the case actually
`pkg/joinserver` should not import `pkg/joinserver/redis`, but `pkg/joinserver/redis` should be able to do the reverse.
In any case, `pkg/joinserver/internal` is also an option.

**Pull request merged**